### PR TITLE
feat(gas_oracle): Add Lru Cache for block values of recent blocks

### LINKED ISSUE #5052: Cache block values of recent blocks
### Describe the feature

when we compute the suggested tip cap, we're searching for the lowest effective tip values of recent blocks. This set is fixed and can be cached

## TODO
* add LRU cache of `Hash -> (Hash, Vec<U256>)` that caches the outcome of: https://github.com/paradigmxyz/reth/blob/b399e5d55232b5bd3a88fc8da9f3681d62519919/crates/rpc/rpc/src/eth/gas_oracle.rs#L201-L202

The cache only needs `max_block` entries if there's activity but there can be more if there aren't any transactions, perhaps a reasonable amount would be 100 entries, 

https://github.com/paradigmxyz/reth/blob/b399e5d55232b5bd3a88fc8da9f3681d62519919/crates/rpc/rpc/src/eth/gas_oracle.rs#L26-L27

the cache needs to be moved into the mutex:

https://github.com/paradigmxyz/reth/blob/b399e5d55232b5bd3a88fc8da9f3681d62519919/crates/rpc/rpc/src/eth/gas_oracle.rs#L93-L94

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PatStiles**
Closes #5052. Would appreciate any pointers to eliminate the unnecessary clone and .to_owned().

### CODE REVIEW

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 1]
```diff
@@ -79,6 +85,34 @@ impl GasPriceOracleConfig {
...
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for EffectiveTipLruCache {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
```
Comment (rkrasiuk) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:105]: nit:
```suggestion
use derive_more::{Deref, DerefMut};

/// Wrapper struct for LruMap
#[derive(Deref, DerefMut))
pub struct EffectiveTipLruCache(LruMap<BlockHash, (BlockHash, Vec<U256>), ByLength>);
```

Comment (PatStiles) [reply crates/rpc/rpc/src/eth/gas_oracle.rs:105]: Thanks for the feeback!

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 2]
```diff
@@ -92,6 +111,8 @@ pub struct GasPriceOracle<Provider> {
     ignore_price: Option<u128>,
     /// The latest calculated price and its block hash
     last_price: Mutex<GasPriceOracleResult>,
+    /// Cache storing the lowest effective tip values of recent blocks
+    lowest_effective_tip_cache: Mutex<EffectiveTipLruCache>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:115]: these should be combined into a single type.

let's call it `GasPriceOracleInner { last_price, lowest_effective_tip_cache}`

File: crates/rpc/rpc/src/eth/gas_oracle.rs [hunk 3]
```diff
@@ -79,6 +83,21 @@ impl GasPriceOracleConfig {
...
+pub struct EffectiveTipLruCache(LruMap<BlockHash, (BlockHash, Vec<U256>), ByLength>);
+
+impl Debug for EffectiveTipLruCache {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("EffectiveTipLruCache")
+            .field("cache_length", &self.len())
+            .field("cache_memory_usage", &self.memory_usage())
+            .finish()
+    }
+}
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/gas_oracle.rs:99]: if we add a new type `GasPriceOracleInner` then we don't need this wrapper and can just use the LruMap as a field directly

**Review by mattsse — CHANGES_REQUESTED** [review]
this does exactly what we need.

some nits and suggestions

File: crates/rpc/rpc/src/eth/gas_oracle.rs (hunk 3 above)
Comment (PatStiles) [reply crates/rpc/rpc/src/eth/gas_oracle.rs:99]: LruMap doesn't implement Debug natively. Do have any suggestions besides adding a wrapper as syntactic sugar to work around having to implement Debug for LruMap<> when its an external crate.

Comment (mattsse) [reply crates/rpc/rpc/src/eth/gas_oracle.rs]: can impl debug manually for GasPriceOracleInner

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**